File: 04_OO/ObjCredit.c
#include <stdio.h>

typedef struct {
    char name[50];
    int credits;  // 取得単位数
} Student;

void addCredits(Student *s, int amount) {   // 単位の加算
    s->credits += amount;
}

void resetCredits(Student *s) {     // 単位をリセット
    s->credits = 0;
}

int main (void) {
    Student alice = {"Alice", 30};

    addCredits(&alice, 10); // OK
    alice.credits = -5;     // 不正な値の直接代入
    alice.credits = 999;    // 不正な値の直接代入
    resetCredits(&alice);   // 2単位引くのではなく、一旦リセット
    alice.credits = 8;      // 　してから合計単位を直接代入

    return 0;
}

File: 04_OO/ObjCredit.java
class Student {
    private String name;
    private int credits;  // private で保護

    public Student(String name) {
        this.name = name;
        this.credits = 0;
    }

    // 単位加算：バリデーション付き
    public void addCredits(int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("不正な単位数：<0");
        }
        if (credits + amount > 200) {  // 上限チェック
            throw new IllegalArgumentException("取得上限を超えています");
        }
        credits += amount;
    }

    public int getCredits() {
        return credits;
    }
}

public class ObjCredit {
    public static void main(String[] args) {
        Student alice = new Student("Alice");

        alice.addCredits(30);  // OK
        System.out.println(alice.getCredits());  // 30

        // alice.credits = -5;  // 直接アクセス不可
    }
}

